File: 3761-minimum-absolute-distance-between-mirror-pairs/3761-minimum-absolute-distance-between-mirror-pairs.go
func minMirrorPairDistance(nums []int) int {
    prev := map[int]int{}
    ans := 1_000_000_007
    for i, num := range nums {
        if _, ok := prev[num]; ok {
            ans = min(ans, i - prev[num])
        }
        
        prev[reverse(num)] = i
    }

    if ans == 1_000_000_007 {
        return -1
    }
    return ans
}

func reverse(n int) int {
	res := 0
	for n > 0 {
		d := n % 10
		res = (res * 10) + d
        n /= 10
	}
    return res
}

// array, hash map
// time: O(n)
// space: O(n)

File: 3761-minimum-absolute-distance-between-mirror-pairs/3761-minimum-absolute-distance-between-mirror-pairs.py
class Solution:
    def minMirrorPairDistance(self, nums: List[int]) -> int:
        prev = dict()
        ans = 1_000_000_007
        for i, num in enumerate(nums):
            if num in prev:
                ans = min(ans, i - prev[num])
            prev[int(str(num)[::-1])] = i  # prev[reverse(num)]
        return -1 if ans == 1_000_000_007 else ans

# array, hash map
# time: O(n)
# space: O(n)